glotaran/model/item.py

```diff
@@ -1,393 +1,686 @@
+from inspect import getmro
+from inspect import isclass
 from textwrap import indent
+from types import NoneType
+from types import UnionType
 from typing import TYPE_CHECKING
+from typing import Any
 from typing import Callable
-from typing import List
+from typing import ClassVar
+from typing import Generator
+from typing import Iterator
 from typing import Type
-
-from glotaran.model.property import ModelProperty
-from glotaran.model.util import wrap_func_as_method
+from typing import TypeAlias
+from typing import TypeVar
+from typing import Union
+from typing import get_args
+from typing import get_origin
+
+from attrs import NOTHING
+from attrs import Attribute
+from attrs import define
+from attrs import evolve
+from attrs import field
+from attrs import fields
+from attrs import resolve_types
+
+from glotaran.parameter import Parameter
+from glotaran.parameter import Parameters
 from glotaran.utils.ipython import MarkdownStr
 
 if TYPE_CHECKING:
-    from typing import Any
-
     from glotaran.model.model import Model
-    from glotaran.parameter import ParameterGroup
 
-    Validator = Callable[
-        [Type[object], Type[Model]],
-        List[str],
-    ]
 
-    ValidatorParameter = Callable[
-        [Type[object], Type[Model], Type[ParameterGroup]],
-        List[str],
-    ]
+META_ALIAS = "__glotaran_alias__"
+META_VALIDATOR = "__glotaran_validator__"
+
+
+class ItemIssue:
+    """Baseclass for item issues."""
+
+    def to_string(self) -> str:
+        """Get the issue as string.
+
+        Returns
+        -------
+        str
+
+        .. # noqa: DAR202
+        .. # noqa: DAR401
+        """
+        raise NotImplementedError
+
+    def __rep__(self) -> str:
+        """Get the representation."""
+        return self.to_string()
+
+
+class ModelItemIssue(ItemIssue):
+    """Issue for missing model items."""
+
+    def __init__(self, item_name: str, label: str):
+        """Create a model issue.
+
+        Parameters
+        ----------
+        item_name : str
+            The name of the item.
+        label : str
+            The item label.
+        """
+        self._item_name = item_name
+        self._label = label
+
+    def to_string(self) -> str:
+        """Get the issue as string.
+
+        Returns
+        -------
+        str
+        """
+        return f"Missing model item '{self._item_name}' with label '{self._label}'."
+
+
+class ParameterIssue(ItemIssue):
+    """Issue for missing parameters."""
+
+    def __init__(self, label: str):
+        """Create a parameter issue.
+
+        Parameters
+        ----------
+        label : str
+            The parameter label.
+        """
+        self._label = label
+
+    def to_string(self) -> str:
+        """Get the issue as string.
+
+        Returns
+        -------
+        str
+        """
+        return f"Missing parameter with label '{self._label}'."
+
+
+class Item:
+    """A baseclass for items."""
+
+    pass
+
+
+@define(kw_only=True, slots=False)
```

Comment:
Consider `slots=False` only where needed, since slots give some performance boot but mess with inheritance and composability.

Reply:
Turns out it in most cases we need  `slots=False`, so it is not worth fine tuning this for the few items where it works with `True`.